### ISSUE TITLE
No balance found after a dry-run shielded IBC with disposable gas payer

### ISSUE THREAD

**Issue by Rigorously** [anoma/namada#2750]
TL;DR: after a specific type of dry-run transaction it is necessary to shielded-sync. Ideally the chain nor the client state should be affected after a dry-run.

First prepare:
```
$ namadac shielded-sync
Syncing finished

$ namadac balance -- owner aspendingkey
Last committed epoch: 21
naan : 33.028
```

Now we run the following admittedly slightly awkward contraption.

```
$ namadac ibc-transfer --token naan --amount 1 --source aspendingkey --receiver the1quickbrownfox --channel-id channel-420 --disposable-gas-payer --gas-spending-key aspendingkey --dry-run-wrapper (or --dry-run)
``` 

I know, it is kind of a ridiculous transaction, but hear me out.

The result is a complaint that the Internal MASP balance is empty.

```
Enter your decryption password: 
Error: 
   0: The balance of the source tnam1pcqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqzmefah is lower than the amount to be transferred. Amount to transfer is 1.000000 tnam1qxvg64psvhwumv3mwrrjfcz0h3t3274hwggyzcee
   1: The balance of the source tnam1pcqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqzmefah is lower than the amount to be transferred. Amount to transfer is 1.000000 tnam1qxvg64psvhwumv3mwrrjfcz0h3t3274hwggyzcee
```

Remember this was a dry-run, so the transaction should have no permanent effect.

```
$ namadac balance --owner aspendingkey
Last committed epoch: 21
No shielded balance found for given key
```

No balance? That is strange!

We sync again and check the balance afterwards.

```
$ namadac shielded-sync
Syncing finished

$ namadac balance -- owner aspendingkey
Last committed epoch: 21
naan : 33.028
```

Indeed, nothing permanent happened on-chain. I guess the client was suffering from impermanent loss. ;)

**Comment by grarco** [comment]
Hi @Rigorously, thanks for bringing this up! It looks indeed like we call `pre_cache_transaction` all the times when generating a masp tx which causes a (possible) update of the context. We should avoid calling it in a dry run
